### PR TITLE
DAOS-2864 control: Refactor daos_server command handling

### PR DESCRIPTION
**Pull request by mjmac**
Pulls command-line parsing out of server package and into daos_server main.
Encourages cleaner designs with clear separation of concerns. Also begins
to eliminate use of global variables by using logger injection, etc.

Adds a new "start" subcommand to daos_server, which means that the user
must specify a mode rather than defaulting to starting the server if no
subcommand was supplied. This makes functionality exploration via CLI help
easier by limiting the number of top-level arguments displayed (i.e. avoid displaying
server arguments for storage subcommands, etc.)

Note to reviewers: This turned into a MASSIVE PR, as it became apparent that the best way to provide a populated configuration to subcommands was to decouple command-line parsing from the configuration itself. 

This PR is only the beginning. There is a lot more work required in order to modularize the codebase. One issue in particular that this PR exposes is that the server configuration struct is passed around way too much, with the result that any changes to it have large ripple effects. It should be decomposed into smaller configuration units which are specific to the code being called.

### CODE REVIEW

File: src/control/cmd/daos_server/start_test.go [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+			},
+			logFnName: "Debug",
+			input:     "hello",
+			wantRe:    regexp.MustCompile(`hello`),
+		},
+		"Error": {
+			configFn: func(cfg *server.Configuration) *server.Configuration {
+				return cfg.WithControlLogMask(server.ControlLogLevelError)
+			},
+			logFnName: "Info",
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/cmd/daos_server/start_test.go]: This is to test that a message at Info level doesn't get emitted if the log level is error. The logging package has a full set of tests -- this is just for setting the log level via config file.

**Comment by mjmac** [comment]
> 
> 
> Didn't find any problems with the actual refactoring, but it's tough to review such a big PR and figure out what's new vs. what's just moved. IMO the change to add the "start" command line parameter should be split into a separate PR if possible.

Yeah, it's a big PR. I considered trying to split it into smaller chunks but I think it's easier to see the whole picture this way. The goal of this and a couple of follow-on PRs is to compartmentalize more of this stuff so that changes don't cut across so many files and package boundaries.

Thanks for being the first to venture into it, though.

File: src/control/cmd/daos_server/start_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/cmd/daos_server/start_test.go]: Ah, got it. Thanks.

**Comment by mjmac** [comment]
> 
> 
> there are too many things going on in this PR, please can we split it up. modifying the usage of daos_server to require a subcommand definitely needs to be split into its own PR as it's going to affect everyone in the team (and will need to be accompanied by e-mail to mailing lists). please could each area of change be identified in a separate PR so that it can be reviewed effectively.

There are multiple commits in this PR; when you look at the "Files changed" view, it's possible to adjust scope from "Changes from all commits" to specific commits. It was neater before I had to merge in master a few times -- there were 4 commits (1 for adding go-cmp, 1 for the daos_server changes, 1 for test harness updates, and 1 for doc updates). Pulling them out into individual PRs will make it harder to review, IMO, as the commits are all logically related.

What if I clean things up again and force-push to get back to the 4 commits?

**Comment by tanabarr** [comment]
> > there are too many things going on in this PR, please can we split it up. modifying the usage of daos_server to require a subcommand definitely needs to be split into its own PR as it's going to affect everyone in the team (and will need to be accompanied by e-mail to mailing lists). please could each area of change be identified in a separate PR so that it can be reviewed effectively.
> 
> There are multiple commits in this PR; when you look at the "Files changed" view, it's possible to adjust scope from "Changes from all commits" to specific commits. It was neater before I had to merge in master a few times -- there were 4 commits (1 for adding go-cmp, 1 for the daos_server changes, 1 for test harness updates, and 1 for doc updates). Pulling them out into individual PRs will make it harder to review, IMO, as the commits are all logically related.
> 
> What if I clean things up again and force-push to get back to the 4 commits?

yes that sounds good

File: src/control/cmd/daos_server/start.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+			WithDebugLogger(logging.NewDebugLogger(f))
+		applyLogConfig()
+
+		return nil
+	}
+
+	cmd.log.Info("no control log file specified; logging to stdout")
+	applyLogConfig()
+
+	for i, srv := range cmd.config.Servers {
```
Comment (tanabarr) [inline src/control/cmd/daos_server/start.go]: this check is not performed if ControlLogFile is specified, whereas it should be regardless of whether the control log file is specified

Comment (mjmac) [reply src/control/cmd/daos_server/start.go]: This function was moved mostly-unchanged from server/config.go: https://github.com/daos-stack/daos/blob/master/src/control/server/config.go#L341

If this is an existing bug to be fixed, let's create a ticket for it. The ioserver configuration stuff is going to be changed in the near future anyhow (follow-on work to this).

**Comment by jbrosenz** [comment]
It is a large set of files to review, but rather than make it fewer files to review, how about a code walk-through to educate us on the major changes?

File: src/control/cmd/daos_server/storage.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+		} else {
+			cmd.log.Infof("NVMe: %s", str)
+		}
+	}
+
+	modules, err := srv.ScanSCM()
+	if err != nil {
+		scanErrors = append(scanErrors, err)
+	} else {
+		str, err := common.StructsToString(modules)
```
Comment (tanabarr) [inline src/control/cmd/daos_server/storage.go]: as above

File: src/control/cmd/daos_server/storage.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+		return errors.New("subcommand must be run as root or sudo")
+	}
+
+	srv, err := server.NewControlService(cmd.config)
+	if err != nil {
+		return errors.WithMessage(err, "initialising ControlService")
+	}
+
+	rebootStr, devices, err := srv.PrepScm(server.PrepScmRequest{
+		Reset: cmd.Reset,
```
Comment (tanabarr) [inline src/control/cmd/daos_server/storage.go]: this kind of changing behaviour should really be in its own PR IMO

File: src/control/cmd/daos_server/start.go (hunk 2 above)
Comment (tanabarr) [reply src/control/cmd/daos_server/start.go]: it's due to the change you applied, returning after checking for the control log file. I don't think there was a return before

File: src/control/cmd/daos_server/start.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		if srv.LogFile == "" {
+			cmd.log.Errorf("no daos log file specified for server %d", i)
+		}
+	}
+
+	return nil
+}
+
+func (cmd *startCmd) Execute(args []string) error {
+	if cmd.start == nil {
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/cmd/daos_server/start.go]: Makes unit testing easier. There's a lot of functionality to test without actually starting anything up.

File: src/control/cmd/daos_server/storage.go (hunk 4 above)
Comment (mjmac) [reply src/control/cmd/daos_server/storage.go]: When these commands lived in the server package, they could use private struct members in the previously-unexported control service. I had a choice here between making those struct members public or expanding the control service API to hide some of the complexity. As I don't think callers really should know about the details I chose the latter approach.

File: src/control/server/iosrv.go [hunk 6]
```diff
@@ -359,8 +381,16 @@ func (srv *iosrv) startCmd() error {
...
-	srv.cmd.Stderr = os.Stderr
+	srv.cmd.Stdout = &cmdLogger{
+		logFn:  srv.log.Info,
+		prefix: fmt.Sprintf("%s:%d", ioServerBin, srv.index),
+	}
+	srv.cmd.Stderr = &cmdLogger{
+		logFn:  srv.log.Error,
+		prefix: fmt.Sprintf("%s:%d", ioServerBin, srv.index),
+	}
+	// FIXME(DAOS-3105): This shouldn't be the default. The command environment
```
Comment (tanabarr) [inline src/control/server/iosrv.go]: I thought this in the past and still agree but unfortunately it's not trivial due to some of the dependencies of the IO server. needs investigation.

File: src/control/cmd/daos_server/start.go (hunk 2 above)
Comment (tanabarr) [reply src/control/cmd/daos_server/start.go]: address in https://github.com/daos-stack/daos/pull/981

File: src/control/cmd/daos_server/storage.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	})
+	if err != nil {
+		return err
+	}
+
+	// Nothing to do after reset
+	if cmd.Reset {
+		return nil
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/cmd/daos_server/storage.go]: addressed in https://github.com/daos-stack/daos/pull/981

File: src/control/cmd/daos_server/start.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/cmd/daos_server/start.go]: addressed in https://github.com/daos-stack/daos/pull/981

**Comment by tanabarr** [comment]
copyright notices also updated in https://github.com/daos-stack/daos/pull/981

File: src/control/cmd/daos_server/main.go [hunk 8]
```diff
@@ -1,17 +1,113 @@
...
+
+func exitWithError(log *logging.LeveledLogger, err error) {
+	log.Debugf("%+v", err)
+	log.Errorf("%v", err)
+	os.Exit(1)
+}
+
+func parseOpts(args []string, opts *mainOpts, log *logging.LeveledLogger) error {
+	p := flags.NewParser(opts, flags.HelpFlag|flags.PassDoubleDash)
+	// TODO(DAOS-3129): Remove this when subcommands are required, in order
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/cmd/daos_server/main.go]: Yes, that's correct.

File: src/control/server/iosrv.go (hunk 6 above)
Comment (mjmac) [reply src/control/server/iosrv.go]: Agreed, which is why I made it a FIXME with a ticket so that it doesn't get lost. I think that longer term we should not support passing configuration through external environment because it introduces a great deal of uncertainty about how things were actually configured, making support more difficult.

**Comment by mjmac** [comment]
> 
> 
> copyright notices also updated in #981

Thank you -- I'm annoyed that I managed to forget them. Will look into ways to flag these via commit hook and/or CI.
